src/riak_kv_query_buffer.erl

```diff
@@ -0,0 +1,599 @@
+        agg = InitAgg,
+        type = Type
+    }.
+
+min_buffer() -> ?MIN_BUFFER.
+
+-spec add(binary()|{binary(), binary()}, buffer()) -> buffer().
+add(Key, #buffer{max_size = MS, count = MS} = Buffer) ->
+    add(Key, merge(Buffer));
+add(_Key,
+    #buffer{
+        rawkey_acc = none, key_acc = none, term_acc = none, rawterm_acc = none,
+        count = C
+    } = B) ->
+    B#buffer{count = C + 1};
+add(Key, #buffer{rawkey_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Key) ->
+    Buffer#buffer{rawkey_acc = [Key|A], count = C + 1};
+add(Key, #buffer{key_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Key) ->
+    Buffer#buffer{key_acc = [{Key}|A], count = C + 1};
+add({Term, Key}, #buffer{rawterm_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Term), is_binary(Key) ->
+    Buffer#buffer{rawterm_acc = [{Term, Key}|A], count = C + 1};
+add({Term, Key}, #buffer{term_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Term), is_binary(Key) ->
+    Buffer#buffer{term_acc = [{{Term, Key}}|A], count = C + 1}.
+
+-spec merge(buffer()) -> buffer().
+merge(#buffer{count = C, key_acc = Acc, type = T} = Buffer)
+        when Acc == none, T == raw_count ->
+    reply(Buffer, {T, C}),
+    Buffer#buffer{count = 0};
+merge(#buffer{key_acc = Acc, agg = Agg = Agg, type = T} = Buffer)
+        when 
+            Acc =/= none, Agg == none, T == keys ->
+    reply(Buffer, {T, Acc}),
+    Buffer#buffer{key_acc = [], count = 0};
+merge(#buffer{rawkey_acc = Acc, agg = Agg = Agg, type = T} = Buffer)
+        when 
+            Acc =/= none, Agg == none, T == raw_keys ->
+    reply(Buffer, {T, Acc}),
+    Buffer#buffer{rawkey_acc = [], count = 0};
+merge(#buffer{term_acc = Acc, agg = Agg = Agg, type = T} = Buffer)
+        when 
+            Acc =/= none , Agg == none, T == terms ->
+    reply(Buffer, {T, Acc}),
+    Buffer#buffer{term_acc = [], count = 0};
+merge(#buffer{rawterm_acc = Acc, agg = Agg = Agg, type = T} = Buffer)
+        when 
+            Acc =/= none, Agg == none, T == raw_terms ->
+    reply(Buffer, {T, Acc}),
+    Buffer#buffer{rawterm_acc = [], count = 0};
+merge(#buffer{rawkey_acc = Acc, agg = Agg, type = T} = Buffer)
+        when Acc =/= none, is_record(Agg, key_agg), T == count  ->
+    reply(Buffer, ping),
+    Buffer#buffer{
+        count = 0,
+        rawkey_acc = [],
+        agg =
+            #key_agg{
+                set = 
+                    sets:union(
+                        sets:from_list(Acc, ?VERSION),
+                        Agg#key_agg.set
+                    )
+            }
+    };
+merge(
+    #buffer{
+        term_acc = Acc, agg = #termcount_agg{map = TCMap}, type = T} = Buffer
+    )
+        when
+            Acc =/= none,
+            is_map(TCMap),
+            T == term_with_rawcount ->
+    reply(Buffer, ping),
+    UpdTCMap =
+        lists:foldl(
+            fun({{Term, _Key}}, FoldAcc) when is_map(FoldAcc), is_binary(Term) ->
```

Comment:
Here is really does not matter whether there is a `{{}}` or just one level of `{}`?

Reply:
See previous comment